pool/paymentmgr.go

```diff
@@ -595,71 +644,237 @@ func (pm *PaymentMgr) maturePendingPayments(height uint32) (map[uint32][]*Paymen
+			if pmt.Account == poolFeesK {
+				_, ok := outputs[feeAddr.String()]
+				if !ok {
+					outputs[feeAddr.String()] = pmt.Amount
+					tOut += pmt.Amount
+					continue
+				}
+				outputs[feeAddr.String()] += pmt.Amount
+				tOut += pmt.Amount
+				continue
+			}
+
+			acc, err := FetchAccount(pm.cfg.DB, []byte(pmt.Account))
+			if err != nil {
+				return err
+			}
+
+			_, ok := outputs[acc.Address]
+			if !ok {
+				outputs[acc.Address] = pmt.Amount
+				tOut += pmt.Amount
+				continue
+			}
+			outputs[acc.Address] += pmt.Amount
+			tOut += pmt.Amount
+		}
+	}
+
+	var diff dcrutil.Amount
+	if tOut > tIn {
+		diff = tOut - tIn
+	} else {
+		diff = tIn - tOut
+	}
+
+	// Ensure the generated outputs are equal to the inputs being spent.
+	if diff != zeroAmount && diff > maxValueDifference {
+		return fmt.Errorf("total input %s and output %s value mismatch",
+			tIn, tOut)
+	}
+
+	inSizes := make([]int, len(inputs))
+	for range inputs {
+		inSizes = append(inSizes, txsizes.RedeemP2PKHSigScriptSize)
+	}
+
+	outSizes := make([]int, len(outputs))
+	for range outputs {
+		inSizes = append(inSizes, txsizes.P2PKHOutputSize)
 	}
 
-	txid, err := pm.cfg.PublishTransaction(pmts, *targetAmt)
+	estSize := txsizes.EstimateSerializeSizeFromScriptSizes(inSizes, outSizes,
+		txsizes.P2PKHOutputSize)
+	estFee := txrules.FeeForSerializeSize(txrules.DefaultRelayFeePerKb, estSize)
+
+	// Deduct the transaction fees from the pool fees being paid out.
+	fees := outputs[feeAddr.String()]
+	fees = fees - estFee
+
+	// Deduct any percentage rounding value differences from the pool
+	// fees being paid out.
+	if diff != zeroAmount {
+		fees = fees - diff
+	}
+
+	outputs[feeAddr.String()] = fees
+
+	// Generate the output set with decoded addresses.
+	outs := make(map[dcrutil.Address]dcrutil.Amount, len(outputs))
+	for sAddr, amt := range outputs {
+		addr, err := dcrutil.DecodeAddress(sAddr, pm.cfg.ActiveNet)
+		if err != nil {
+			return fmt.Errorf("unable to decode address: %v", err)
+		}
+
+		outs[addr] = amt
+	}
+
+	tx, err := txCreator.CreateRawTransaction(inputs, outs, nil, nil)
+	if err != nil {
+		return fmt.Errorf("unable to create raw transaction: %v", err)
+	}
+
+	txB, err := tx.Bytes()
 	if err != nil {
 		return err
 	}
-	for _, bundle := range eligiblePmts {
-		bundle.UpdateAsPaid(pm.cfg.DB, height, txid)
-		err = bundle.ArchivePayments(pm.cfg.DB)
+
+	for {
+		// Ensure the wallet is synced to the current chain height before
+		// proceeding with payments.
+		walletHeight, err := pm.cfg.WalletBestHeight()
 		if err != nil {
-			return err
+			return fmt.Errorf("unable to fetch best wallet height: %v", err)
+		}
+
+		if height == walletHeight {
+			break
 		}
 	}
-	err = pm.cfg.DB.Update(func(tx *bolt.Tx) error {
-		err = pm.persistTxFeeReserve(tx)
-		if err != nil {
-			return err
+
+	txBroadcaster := pm.cfg.FetchTxBroadcaster()
+	if txBroadcaster == nil {
+		return fmt.Errorf("tx broadcastor unset")
+	}
+
+	balanceReq := &walletrpc.BalanceRequest{
```

Comment:
Would like to see a comment here, something like 

```
// Balance request not technically required here, just used so we can provide a helpful error message. 
```

Reply:
I'd say it is required there, the spendable prev outs from the wallet cannot be less than the total outputs paying from those prev outs.